contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract EtomicSwapV2 {
+    enum MakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerSpent,
+        MakerRefunded
+    }
+
+    struct MakerPayment {
+        bytes20 paymentHash;
+        uint32 paymentLockTime;
+        MakerPaymentState state;
+    }
+
+    event MakerPaymentSent(bytes32 id);
+    event MakerPaymentSpent(bytes32 id, bytes32 secret);
+    event MakerPaymentRefundedTimelock(bytes32 id);
+    event MakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+
+    mapping(bytes32 => MakerPayment) public makerPayments;
+
+    enum TakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerApproved,
+        MakerSpent,
+        TakerRefunded
+    }
+
+    struct TakerPayment {
+        bytes20 paymentHash;
+        uint32 immediateRefundTime;
+        uint32 paymentLockTime;
+        TakerPaymentState state;
+    }
+
+    event TakerPaymentSent(bytes32 id);
+    event TakerPaymentApproved(bytes32 id);
+    event TakerPaymentSpent(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedTimelock(bytes32 id);
+
+    mapping(bytes32 => TakerPayment) public takerPayments;
+
+    address public immutable dexFeeAddress;
+
+    constructor(address feeAddress) {
+        require(feeAddress != address(0), "feeAddress must not be zero address");
+
+        dexFeeAddress = feeAddress;
+    }
+
+    function ethMakerPayment(
+        bytes32 id,
```

Comment:
Why not use `ripemd160` for the id to reduce storage, I guess hash/id collision will still be extremely low with a 20 bytes id.

Reply:
Similar to above, the `id` will be padded with zeros and stored in 32 bytes slot, so there is a negligible difference of gas used:
1. With 32 bytes id: 90713 gas.
2. With 20 bytes id: 90693 gas.

Pushed another demo commit: https://github.com/KomodoPlatform/etomic-swap/commit/79ebeea8c832af38476895de4c70f78af5f220eb